File: back-end/src/main/java/es/ull/project/domain/valueobject/cost/TotalCost.java
package es.ull.project.domain.valueobject.cost;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Objects;
 

/**
 * TotalCost
 * 
 * Represents the total cost in the domain, which can include
 * opening fixed costs, transportation variable costs, or other cost components.
 * This is a mandatory value object in the domain.
 */
public final class TotalCost {

    private static final int ZERO = 0;
    private static final String ERROR_AMOUNT_NOT_DEFINED = "Total cost is not defined";
    private static final String ERROR_AMOUNT_NEGATIVE = "Total cost cannot be negative";
    private static final String ERROR_CURRENCY_INVALID = "Currency cannot be null or empty";
    private static final String ERROR_CURRENCY_MISMATCH = "Cannot operate on costs with different currencies";

    /**
     * Amount of the total cost.
     * Always stored with 2 decimal precision.
     * It is a required attribute.
     */
    private final BigDecimal amount;

    /**
     * Currency of the cost. Default: "EUR".
     * It is an optional attribute.
     */
    private final Currency currency;

    /**
     * Creates a new TotalCost with a default currency (EUR).
     *
     * @param amount Amount of the total cost. Must be ≥ 0.
     */
    public TotalCost(double amount) {
        this(amount, new Currency());
    }

    /**
     * Creates a new TotalCost with a specific currency. ()BETTER)
     *
     * @param amount   Amount of the total cost. Must be ≥ 0.
     * @param currency Currency code (e.g., "EUR", "USD"). Cannot be null/empty.
     */
    public TotalCost(double amount, Currency currency) {
        validateAmount(amount);
        validateCurrency(currency);
        this.amount = BigDecimal.valueOf(amount).setScale(2, RoundingMode.HALF_UP);
        this.currency = currency;
    }

    /**
     * Creates a new TotalCost with a specific currency.
     *
     * @param amount   Amount of the total cost. Must be ≥ 0.
     * @param currency Currency code (e.g., "EUR", "USD"). Cannot be null/empty. It creates a Currency object internally.
     */
    public TotalCost(double amount, String currency) {
        validateAmount(amount);
        validateCurrencyString(currency);
        this.amount = BigDecimal.valueOf(amount).setScale(2, RoundingMode.HALF_UP);
        this.currency = new Currency(currency);
    }

    /**
     * Validates that the amount is defined and non-negative.
     *
     * @param amount Amount to validate.
     * @throws IllegalArgumentException if amount is NaN or negative.
     */
    private void validateAmount(double amount) {
        if (Double.isNaN(amount)) {
            throw new IllegalArgumentException(ERROR_AMOUNT_NOT_DEFINED);
        }
        if (amount < ZERO) {
            throw new IllegalArgumentException(ERROR_AMOUNT_NEGATIVE);
        }
    }

    /**
     * Validates that the currency string is not null or empty.
     *
     * @param currency Currency string to validate.
     * @throws IllegalArgumentException if currency is null or empty.
     */
    private void validateCurrencyString(String currency) {
        if (currency == null || currency.isEmpty()) {
            throw new IllegalArgumentException(ERROR_CURRENCY_INVALID);
        }
    }

    /**
     * Validates that the currency object is not null.
     *
     * @param currency Currency object to validate.
     * @throws IllegalArgumentException if currency is null.
     */
    private void validateCurrency(Currency currency) {
        if (currency == null) {
            throw new IllegalArgumentException(ERROR_CURRENCY_INVALID);
        }
    }

    /**
     * Returns the amount of the total cost.
     *
     * @return Amount as a double value.
     */
    public double getAmount() {
        return this.amount.doubleValue();
    }

    /**
     * Returns the currency of the total cost.
     *
     * @return Currency object (never null)
     */
    public Currency getCurrency() {
        return this.currency;
    }

    /**
     * Returns a new TotalCost with the sum of this and another cost.
     *
     * @param other Other TotalCost to add.
     * @return New TotalCost representing the sum.
     */
    public TotalCost add(TotalCost other) {
        checkCurrencyCompatibility(other);
        return new TotalCost(this.amount.add(other.amount).doubleValue(), this.currency);
    }

    /**
     * Returns a new TotalCost with the difference of this and another cost.
     *
     * @param other Other TotalCost to subtract.
     * @return New TotalCost representing the difference.
     */
    public TotalCost subtract(TotalCost other) {
        checkCurrencyCompatibility(other);
        double result = this.amount.subtract(other.amount).doubleValue();
        if (result < ZERO) {
            result = ZERO;
        }
        return new TotalCost(result, this.currency);
    }

    /**
     * Compares this cost to another to check if greater.
     *
     * @param other Other TotalCost.
     * @return True if this cost > other.
     */
    public boolean greaterThan(TotalCost other) {
        checkCurrencyCompatibility(other);
        return this.amount.compareTo(other.amount) > ZERO;
    }

    /**
     * Compares this cost to a maximum budget to check if greater.
     *
     * @param other MaximumBudget to compare with.
     * @return True if this cost exceeds the budget.
     * @throws IllegalArgumentException if currencies don't match.
     */
    public boolean greaterThan(MaximumBudget other) {
        checkCurrencyCompatibility(other);
        return this.amount.compareTo(BigDecimal.valueOf(other.getAmount())) > ZERO;
    }

    /**
     * Compares this cost to another to check if less.
     *
     * @param other Other TotalCost.
     * @return True if this cost < other.
     */
    public boolean lessThan(TotalCost other) {
        checkCurrencyCompatibility(other);
        return this.amount.compareTo(other.amount) < ZERO;
    }

    /**
     * Compares this cost to another to check if less.
     *
     * @param other Other TotalCost.
     * @return True if this cost < other.
     */
    public boolean lessThan(MaximumBudget other) {
        checkCurrencyCompatibility(other);
        return this.amount.compareTo(BigDecimal.valueOf(other.getAmount())) < ZERO;
    }

    /**
     * Checks if the currency of this cost matches another cost's currency.
     *
     * @param other The total cost to check against.
     * @throws IllegalArgumentException if currencies don't match.
     */
    private void checkCurrencyCompatibility(TotalCost other) {
        if (!this.currency.equals(other.currency)) {
            throw new IllegalArgumentException(ERROR_CURRENCY_MISMATCH);
        }
    }

    /**
     * Checks if the currency of this cost matches a maximum budget's currency.
     *
     * @param other The maximum budget to check against.
     * @throws IllegalArgumentException if currencies don't match.
     */
    private void checkCurrencyCompatibility(MaximumBudget other) {
        if (!this.currency.equals(other.getCurrency())) {
            throw new IllegalArgumentException(ERROR_CURRENCY_MISMATCH);
        }
    }

    /**
     * Checks equality based on amount and currency.
     *
     * @param otherObject Object to compare with.
     * @return True if amounts and currencies are equal, false otherwise.
     */
    @Override
    public boolean equals(Object otherObject) {
        if (this == otherObject) {
            return true;
        }
        if (otherObject == null || getClass() != otherObject.getClass()) {
            return false;
        }
        TotalCost other = (TotalCost) otherObject;
        return this.amount.equals(other.amount) && this.currency.equals(other.currency);
    }

    /**
     * Generates hash code based on amount and currency.
     *
     * @return Hash code value.
     */
    @Override
    public int hashCode() {
        return Objects.hash(amount, currency);
    }

    /**
     * Returns a string representation of the total cost.
     *
     * @return Formatted string with amount and currency.
     */
    @Override
    public String toString() {
        return String.format("TotalCost={amount=%.2f, currency='%s'}", this.amount, this.currency);
    }
}

File: front-end/src/domain/valueobject/cost/total-cost.ts
import { Currency } from './currency';

/**
 * TotalCost
 *
 * Represents the total cost with amount and currency.
 */

export class TotalCost {
  /**
   * Error message when the amount is not defined.
   */
  private static readonly ERROR_AMOUNT_NOT_DEFINED = 'Total cost is not defined';
  /**
   * Error message when the amount is negative.
   */
  private static readonly ERROR_AMOUNT_NEGATIVE = 'Total cost cannot be negative';

  /**
   * Amount of the total cost (stored with 2 decimals).
   */
  readonly amount: number;
  /**
   * Currency of the total cost.
   */
  readonly currency: Currency;

  /**
   * Create a TotalCost.
   * @param amount numeric amount (must be >= 0)
   * @param currency either a Currency instance or ISO code string; defaults to EUR
   */
  constructor(amount: number, currency?: Currency | string) {
    this.validateAmount(amount);
    if (currency == null) {
      this.currency = new Currency();
    } else if (typeof currency === 'string') {
      this.currency = new Currency(currency);
    } else {
      this.currency = currency;
    }
    this.amount = Math.round((amount + Number.EPSILON) * 100) / 100;
  }

  /**
   * Validate the numeric amount.
   * @param amount candidate amount
   * @throws Error when NaN or negative
   */
  private validateAmount(amount: number) {
    if (Number.isNaN(amount)) throw new Error(TotalCost.ERROR_AMOUNT_NOT_DEFINED);
    if (amount < 0) throw new Error(TotalCost.ERROR_AMOUNT_NEGATIVE);
  }


  /**
   * Get amount (2-decimal precision).
   * @returns The numeric amount (2 decimals).
   */
  getAmount(): number { return this.amount; }

  /**
   * Get currency instance.
   * @returns The currency instance.
   */
  getCurrency(): Currency { return this.currency; }


  /**
   * Add another TotalCost with the same currency.
   * @param other Cost to add.
   * @returns A new TotalCost with the sum.
   */
  add(other: TotalCost): TotalCost {
    this.checkCurrencyCompatibility(other);
    return new TotalCost(this.amount + other.amount, this.currency);
  }


  /**
   * Subtract another TotalCost (result floored at 0).
   * @param other Cost to subtract.
   * @returns A new TotalCost with the difference (never negative).
   */
  subtract(other: TotalCost): TotalCost {
    this.checkCurrencyCompatibility(other);
    let result = this.amount - other.amount;
    if (result < 0) result = 0;
    return new TotalCost(result, this.currency);
  }


  /**
   * Compare greater-than with another TotalCost (same currency required).
   * @param other The other TotalCost to compare.
   * @returns True if this amount is greater than the other's amount.
   */
  greaterThan(other: TotalCost): boolean {
    this.checkCurrencyCompatibility(other);
    return this.amount > other.amount;
  }


  /**
   * Compare less-than with another TotalCost (same currency required).
   * @param other The other TotalCost to compare.
   * @returns True if this amount is less than the other's amount.
   */
  lessThan(other: TotalCost): boolean {
    this.checkCurrencyCompatibility(other);
    return this.amount < other.amount;
  }


  /**
   * Ensure both costs use the same currency.
   * @param other The other TotalCost or object with getCurrency().
   * @throws Error when currencies differ
   */
  private checkCurrencyCompatibility(other: TotalCost | { getCurrency(): Currency }) {
    if (!this.currency.equals((other as any).getCurrency())) {
      throw new Error('Cannot operate on costs with different currencies');
    }
  }


  /**
   * Equality with tolerance for floating rounding (0.005).
   * @param other The object to compare.
   * @returns True if both TotalCost are equal (with tolerance).
   */
  equals(other: unknown): boolean {
    if (this === other) return true;
    if (!(other instanceof TotalCost)) return false;
    return Math.abs(this.amount - other.amount) < 0.005 && this.currency.equals(other.currency);
  }

  /**
   * Human-readable representation with two decimals.
   * @returns String representation of the TotalCost.
   */
  toString(): string {
    return `TotalCost={amount=${this.amount.toFixed(2)}, currency='${this.currency.getCode()}'}`;
  }
}
